pkg/kuberneteshelper/pod.go

```diff
@@ -105,3 +108,32 @@ func GetOsmConfigurator(osmNamespace common.MeshNamespace) configurator.Configur
 	cfg := configurator.NewConfigurator(versioned.NewForConfigOrDie(kubeConfig), stop, osmNamespace.String(), constants.OSMMeshConfig)
 	return cfg
 }
+
+// GetMatchingServices returns a list of Kuberentes services in the namespace that match the pod's label
+func GetMatchingServices(kubeClient kubernetes.Interface, podLabels map[string]string, namespace string) ([]service.MeshService, error) {
+	serviceList, err := kubeClient.CoreV1().Services(namespace).List(context.Background(), v12.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	matchingServiceSet := mapset.NewSet()
+	for _, svc := range serviceList.Items {
+		svcRawSelector := svc.Spec.Selector
+		selector := labels.Set(svcRawSelector).AsSelector()
+		// service has no selectors, we do not need to match against the pod label
+		if len(svcRawSelector) == 0 {
+			continue
+		}
+		if selector.Matches(labels.Set(podLabels)) {
+			meshSvc := service.MeshService{
+				Namespace: namespace,
+				Name:      svc.Name,
+			}
+			matchingServiceSet.Add(meshSvc)
+		}
+	}
+	var matchingServiceList []service.MeshService
+	for matchedSvc := range matchingServiceSet.Iter() {
+		matchingServiceList = append(matchingServiceList, matchedSvc.(service.MeshService))
+	}
+	return matchingServiceList, nil
```

Comment:
@SanyaKochhar is there any reason why we can't create a slice and append the results to the slice immediately, instead of creating a set then transferring over to the slice?

Reply:
hm originally this was to ensure no duplicates but the logic has changed now so might be able to do that 🤔